<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>SIB training - Analysis tools and QC</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../assets/SIB_logo.svg" rel="icon" type="image/svg+xml">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato|Sofia|Source+Sans+Pro">


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">SIB training</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-1" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 1</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-1">    
        <li>
    <a class="dropdown-item" href="../day1/analysis_tools_qc.html">
 <span class="dropdown-text">Analysis tools and QC</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Analysis tools and QC</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../assets/SIB_LogoQ_GBv.svg" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#material" id="toc-material" class="nav-link active" data-scroll-target="#material">Material</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises">Exercises</a>
  <ul class="collapse">
  <li><a href="#loading-scrnaseq-data" id="toc-loading-scrnaseq-data" class="nav-link" data-scroll-target="#loading-scrnaseq-data">Loading scRNAseq data</a></li>
  <li><a href="#the-seurat-object" id="toc-the-seurat-object" class="nav-link" data-scroll-target="#the-seurat-object">The <code>Seurat</code> object</a></li>
  <li><a href="#visualizing-qc-per-cell-and-gene" id="toc-visualizing-qc-per-cell-and-gene" class="nav-link" data-scroll-target="#visualizing-qc-per-cell-and-gene">Visualizing QC per cell and gene</a></li>
  <li><a href="#cell-filtering" id="toc-cell-filtering" class="nav-link" data-scroll-target="#cell-filtering">Cell filtering</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Analysis tools and QC</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning outcomes
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Load single cell data into R</li>
<li>Explain the basic structure of a <code>Seurat</code> object and extract count data and metadata</li>
<li>Calculate and visualize quality measures based on:
<ul>
<li>mitochondrial genes</li>
<li>ribosomal genes</li>
<li>hemoglobin genes</li>
<li>relative gene expression</li>
</ul></li>
<li>Interpret the above quality measures per cell.</li>
<li>Perform cell filtering based on quality thresholds</li>
</ul>
</div>
</div>
<section id="material" class="level2">
<h2 class="anchored" data-anchor-id="material">Material</h2>
<p><a href="../assets/pdf/analysis_tools_qc.pdf">Download the presentation</a></p>
</section>
<section id="exercises" class="level2">
<h2 class="anchored" data-anchor-id="exercises">Exercises</h2>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-1_26ec3ddf80da1c6185cc3ce7f7f44723">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://single-cell-transcriptomics.s3.eu-central-1.amazonaws.com/course_data.tar.gz</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> course_data.tar.gz</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> course_data.tar.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="loading-scrnaseq-data" class="level3">
<h3 class="anchored" data-anchor-id="loading-scrnaseq-data">Loading scRNAseq data</h3>
<p>The next step after the generation of the count matrices with <code>cellranger count</code>, is the data analysis. The <code>R</code> package <code>Seurat</code> is currently the most popular software to do this. To start working with <code>Seurat</code> you can load it into your environment like this:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-2_e4a68a4f62618b3702684eee0baa2b4c">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip: make an R script
</div>
</div>
<div class="callout-body-container callout-body">
<p>You could type and copy-paste the commands of these exercises directly in the console. However, that makes it hard to track what you have done. In addition, it can be nice to add comments to your code, so you can read back why you have made certain choices. In order to do that, do not write commands in the console, but write them in a script, and send them to the console with <kbd>^ Ctrl + ⏎ Enter</kbd> (Windows) or <kbd>⌘ Command + ⏎ Enter</kbd> (MacOS).</p>
</div>
</div>
<p>First, we will load a file specifying the different samples, and create an object specifying the location of the count data:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-3_670331fd9ab28ca9ac0d6428e38d965d">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>sample_info <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"course_data/sample_info_course.csv"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>datadirs <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"course_data"</span>, <span class="st">"count_matrices"</span>, sample_info<span class="sc">$</span>SampleName,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"outs"</span>, <span class="st">"filtered_feature_bc_matrix"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(datadirs) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">"-"</span>, sample_info<span class="sc">$</span>SampleName)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>datadirs <span class="ot">&lt;-</span> datadirs[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The object <code>datadirs</code> is a named vector specifying the paths of the count directories for each sample:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-4_9caaa35ad3bf2beb8421e4a7be68804f">
<div class="cell-output cell-output-stdout">
<pre><code>                                                             PBMMC-1 
"course_data/count_matrices/PBMMC_1/outs/filtered_feature_bc_matrix" 
                                                             PBMMC-2 
"course_data/count_matrices/PBMMC_2/outs/filtered_feature_bc_matrix" 
                                                             PBMMC-3 
"course_data/count_matrices/PBMMC_3/outs/filtered_feature_bc_matrix" </code></pre>
</div>
</div>
<p>To run through a typical <code>Seurat</code> analysis, we will use the files that are in the directory <code>data/filtered_feature_bc_matrix</code>. This directory is part of the output generated by <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger"><code>cellranger</code></a>. To load this data into R and generate a sparse matrix, run the following command:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-5_d8c5ea7538d752e34e3bda4313052e2f">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sparse_matrix <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X</span>(<span class="at">data.dir =</span> datadirs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This imports a raw count matrix. Have a look at the counts of the first 30 cells of three genes by running:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-6_f1037ad3ea609b7f33d8e68c000eda4c">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sparse_matrix[<span class="fu">c</span>(<span class="st">"PECAM1"</span>, <span class="st">"CD8A"</span>, <span class="st">"TSPAN1"</span>), <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>3 x 30 sparse Matrix of class "dgCMatrix"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>   [[ suppressing 30 column names 'PBMMC-1_AAACCTGCAGACGCAA-1', 'PBMMC-1_AAACCTGTCATCACCC-1', 'PBMMC-1_AAAGATGCATAAAGGT-1' ... ]]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                                  
PECAM1 . . . . . . . 1 . . . . 1 . . 1 . . . . . . . 1 . . . . . .
CD8A   . 1 . . 1 . . . . . . . . . . . . . . . . . . . . . . . . .
TSPAN1 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .</code></pre>
</div>
</div>
<p>You will see many dots (zeros) and a few integers representing the counts per gene per cell.</p>
<p>To generate a <code>Seurat</code> object, we will run <code>CreateSeuratObject</code>. Note that while creating the object, we already do some mild filtering; we only import genes that are expressed in at least 3 cells, and we only import cells that have at least 100 different genes expressed:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-7_621fa5b005681fb7ad2478c5af46fbea">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>seu <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> sparse_matrix,</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">project =</span> <span class="st">"pbmmc"</span>,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">min.cells =</span> <span class="dv">3</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">min.features =</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Feature names cannot have underscores ('_'), replacing with dashes
('-')</code></pre>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Function notation with <code>::</code>
</div>
</div>
<div class="callout-body-container callout-body">
<p>Here, we define the function together with its associated package. We do that by the syntax <code>package::function</code>. Of course, you can also call <code>library(package)</code>, and only type the function name. Since we use many different packages in this course, it can be confusing which function comes from which package. Therefore, we chose to always associate the package with the called function.</p>
</div>
</div>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>check what’s in the <code>seu</code> object, by typing <code>seu</code> in the R console. How many features are in there? And how many cells?</p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Typing <code>seu</code> should return:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-8_96ed70f96d5223c95ba877065e4ca09b">
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
18673 features across 6946 samples within 1 assay 
Active assay: RNA (18673 features, 0 variable features)</code></pre>
</div>
</div>
<p>This means that we have 18673 genes (features) in there, and 6946 cells (samples)</p>
</div>
</div>
</div>
</section>
<section id="the-seurat-object" class="level3">
<h3 class="anchored" data-anchor-id="the-seurat-object">The <code>Seurat</code> object</h3>
<p>The <code>seu</code> object we have created has the class <code>Seurat</code>. The object contains multi-level slots and lists. Each <code>Seurat</code> object contains exactly the same slots that are specified in the image below (get slot descriptions by typing <code>?SeuratObject::Seurat</code>). You can get the information inside a slot with <code>@</code>, in the same way as you would use the <code>$</code> for lists (e.g.&nbsp;<code>seu@meta.data</code> will return the a data frame with information on each cell). Slots can be filled with other R objects, like lists, vectors, data frames or any other class. Here’s an overview of all slots that are in a <code>Seurat</code> object:</p>
<p><img src="../assets/images/seurat_object.png" class="img-fluid"></p>
<p>In addition to the original count table, the <code>Seurat</code> object can therefore store a lot of information that is generated during your analysis, like results of a normalization (<code>@assays$RNA@data</code>) a PCA or UMAP (<code>@reductions</code>) and the clustering (<code>@graphs</code>). It also tracks all the commands that have been used to generate the object in its current state (<code>@commands</code>). Therefore, while going through the analysis steps, the same object gets more and more of its slots filled. Because most <code>Seurat</code> functions return the input object + adjusted slots, we can use this syntax:</p>
<pre><code>seurat_object &lt;- Seurat::function(seurat_object)</code></pre>
<p>So, the function takes an object as input and we assign it to an object with the same name. Meaning that we overwrite the object used as input. This is fine in many cases, because <code>Seurat</code> adds information to the input object, and returns is. We will use this syntax for the first time at <a href="normalization_scaling.md">Normalization and scaling</a>.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Getting specific information out of the <code>Seurat</code> object
</div>
</div>
<div class="callout-body-container callout-body">
<p>In order to get specific data you can use the <code>@</code> and <code>$</code> symbols to browse through the objects. However, <code>Seurat</code> comes with a lot of convenience functions, that are easier to use. So, e.g.&nbsp;to get the raw count matrix, you could type <code>seurat_object@assays$RNA@counts</code>, however, this is equivalent to <code>GetAssayData(object = seurat_object, slot = "counts")</code>. More information on these convenience functions <a href="https://satijalab.org/seurat/articles/essential_commands.html">here</a>.</p>
</div>
</div>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>A. Have a look at the <code>seu</code> object by running <code>View(seu)</code>. What is in there? What is stored in <code>@active.ident</code>?</p>
<p>B. Have a look at the <code>data.frame</code> stored at <code>seu@meta.data</code> what kind of information is in there?</p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="answer-a" class="level3">
<h3 class="anchored" data-anchor-id="answer-a">Answer A</h3>
<p>There are many slots as described in the above figure. The slot <code>@active.ident</code> contains data specifying the samples, e.g.&nbsp;<code>table(seu@active.ident)</code> returns:</p>
<pre><code>PBMMC-1 PBMMC-2 PBMMC-3 
1612    3105    2229 </code></pre>
<p>Which are the number of cells per sample.</p>
</section>
<section id="answer-b" class="level3">
<h3 class="anchored" data-anchor-id="answer-b">Answer B</h3>
<p>Running <code>head(seu@meta.data)</code> returns:</p>
<pre><code>                            orig.ident nCount_RNA nFeature_RNA
PBMMC-1_AAACCTGCAGACGCAA-1    PBMMC-1       2401          909
PBMMC-1_AAACCTGTCATCACCC-1    PBMMC-1       3532          760
PBMMC-1_AAAGATGCATAAAGGT-1    PBMMC-1       3972         1215
PBMMC-1_AAAGCAAAGCAGCGTA-1    PBMMC-1       3569          894
PBMMC-1_AAAGCAACAATAACGA-1    PBMMC-1       2982          730
PBMMC-1_AAAGCAACATCAGTCA-1    PBMMC-1      22284         3108</code></pre>
<p>Giving you the names of three columns and a row for each cell:</p>
<ul>
<li><code>orig_ident</code>: the original identity (origin) of a cell.</li>
<li><code>nCount_RNA</code>: the number of reads assigned to a cell.</li>
<li><code>nFeature_RNA</code>: the number of expressed features (genes) per cell.</li>
</ul>
</section>
</div>
</div>
</div>
<p>Luckily, usually you do not have to dive into this structure to retrieve information. For example, information in the slot <code>@meta.data</code> can be retrieved and set by using <code>$</code> or <code>[[]]</code>.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>There is a subtle difference here between <code>$</code> and <code>[[]]</code>. While <code>$</code> returns a vector of the column in <code>@meta.data</code>, <code>[[]]</code> returns a <code>data.frame</code>.</p>
</div>
</div>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Generate a histogram of the column <code>nCount_RNA</code> at <code>seu@meta.data</code>, with the base function <code>hist</code>.</p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-9_53c1bca8768838c88270f943cd88df59">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(seu<span class="sc">$</span>nCount_RNA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="analysis_tools_qc_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>or</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-10_4ffdfa0612b63d365a617278190fcf6d">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(seu<span class="sc">@</span>meta.data<span class="sc">$</span>nCount_RNA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="analysis_tools_qc_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
<p>There are also built-in functions to plot data from <code>Seurat</code> object, for example <code>FeatureScatter</code>. This function enables you easily draw a scatterplot from a <code>Seurat</code> object:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-11_ab95d526d9cf0e480e1eedb5327ad5c6">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">FeatureScatter</span>(seu, <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, <span class="at">feature2 =</span> <span class="st">"nFeature_RNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="analysis_tools_qc_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>You can find a nice overview of such functions <a href="https://satijalab.org/seurat/articles/essential_commands.html">here</a>.</p>
</section>
<section id="visualizing-qc-per-cell-and-gene" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-qc-per-cell-and-gene">Visualizing QC per cell and gene</h3>
<p>While generating the <code>Seurat</code> object, there were already some quality measures calculated for each cell, namely the total UMI counts per cell (<code>nCount_RNA</code>) and the total number of detected features per cell (<code>nFeature_RNA</code>). We can plot those in a violin plot and evaluate their distribution per sample:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-12_151b0c1ce9f740ae7415e1fac190b8b5">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">VlnPlot</span>(seu, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>,</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"nFeature_RNA"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="analysis_tools_qc_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>You can see that there is quite a wide range for both. A cell with low number of detected features or counts might not give you a lot of information, while a high number of detected features/counts might point to doublets.</p>
<p>Single cells have often undergone sampling and/or dissociation and/or sorting. Therefore, there are often cells and genes in your dataset that cause variation due to technical reasons. In the following steps, we will visualize those and make decisions on whether or not to remove cells or genes with extreme values.</p>
<p>We will start with calculating the percentage of counts coming from transcript types:</p>
<ul>
<li><strong>Mitochondrial genes</strong>: If a cell membrane is damaged, it looses free RNA quicker compared to mitochondrial RNA, because the latter is part of the mitochondrion. A high relative amount of mitochondrial counts can therefore point to damaged cells (<a href="https://f1000research.com/articles/5-2122">Lun et al.&nbsp;2016</a>).</li>
<li><strong>Ribosomal genes</strong>: Are not rRNA (ribosomal RNA) but is mRNA that code for ribosomal proteins. They do not point to specific issues, but it can be good to have a look at their relative abundance. They can have biological relevance (e.g.&nbsp;<a href="https://www.nature.com/articles/s41598-020-64929-x">Caron et al.&nbsp;2020</a>).</li>
<li><strong>Hemoglobin genes</strong>: these transcripts are very abundant in erythrocytes. Depending on your application, you can expect ‘contamination’ of erythrocytes and select against it.</li>
</ul>
<p>In order to have an idea about the relative counts of these type of genes in our dataset, we can calculate their expression as relative counts in each cell. We do that by selecting genes based on patterns (e.g.&nbsp;<code>^MT-</code> matches with all gene names starting with <code>MT</code>, i.e.&nbsp;mitochondrial genes):</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-13_d28f8d2e840dd13c8d0a89d77ab575ae">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># mitochondrial genes</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>seu <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">PercentageFeatureSet</span>(seu, </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">pattern =</span> <span class="st">"^MT-"</span>, </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">col.name =</span> <span class="st">"percent.mito"</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ribosomal genes</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>seu <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">PercentageFeatureSet</span>(seu, </span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">pattern =</span> <span class="st">"^RP[SL]"</span>,</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">col.name =</span> <span class="st">"percent.ribo"</span>)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="co"># hemoglobin genes (but not HBP)</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>seu <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">PercentageFeatureSet</span>(seu,</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">pattern =</span> <span class="st">"^HB[^(P)]"</span>,</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">col.name =</span> <span class="st">"percent.globin"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Run the commands and check out the metadata data frame at <code>seu@meta.data</code>. What has changed?</p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>If we type <code>head(seu@meta.data)</code> it returns:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-14_9868359ac8c057a6b323f892067f1c07">
<div class="cell-output cell-output-stdout">
<pre><code>                           orig.ident nCount_RNA nFeature_RNA percent.mito
PBMMC-1_AAACCTGCAGACGCAA-1    PBMMC-1       2401          909     2.540608
PBMMC-1_AAACCTGTCATCACCC-1    PBMMC-1       3532          760     5.181200
PBMMC-1_AAAGATGCATAAAGGT-1    PBMMC-1       3972         1215     4.934542
PBMMC-1_AAAGCAAAGCAGCGTA-1    PBMMC-1       3569          894     3.250210
PBMMC-1_AAAGCAACAATAACGA-1    PBMMC-1       2982          730     3.688799
PBMMC-1_AAAGCAACATCAGTCA-1    PBMMC-1      22284         3108     3.181655
                           percent.ribo percent.globin
PBMMC-1_AAACCTGCAGACGCAA-1     28.65473      0.1665973
PBMMC-1_AAACCTGTCATCACCC-1     55.03964      0.1981880
PBMMC-1_AAAGATGCATAAAGGT-1     30.43807      0.3776435
PBMMC-1_AAAGCAAAGCAGCGTA-1     55.02942      0.3642477
PBMMC-1_AAAGCAACAATAACGA-1     54.49363      0.1006036
PBMMC-1_AAAGCAACATCAGTCA-1     23.40693     36.9682283</code></pre>
</div>
</div>
<p>So, the function <code>PercentageFeatureSet</code> adds a column to <code>meta.data</code>, specifying the percentage of counts for the specified gene sets.</p>
</div>
</div>
</div>
<p>Now we can plot the distribution of these percentages in a violin plot:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-15_1766b74660c75827a65995c0142b49f1">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">VlnPlot</span>(seu, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"percent.mito"</span>,</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"percent.ribo"</span>,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"percent.globin"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="analysis_tools_qc_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>You can see that <code>PBMMC-2</code> is quite different from the two others, it has a group of cells with very low ribosomal counts and one with very high globin counts. Maybe these two percentages are negatively correlated? Let’s have a look, by plotting the two percentages against each other:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-16_baf82c8147d1ce80bf8c012a05bfcbd0">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">FeatureScatter</span>(seu, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">feature1 =</span> <span class="st">"percent.globin"</span>, </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">feature2 =</span> <span class="st">"percent.ribo"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="analysis_tools_qc_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Are they correlated? What kind of cells might have a high abundance of hemoglobin transcripts and low ribosomal transcripts?</p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Yes there is a negative correlation. Erythrocytes (red blood cells) have a high abundance of hemoglobin transcripts and low abundance of ribosomal transcripts. These are most likely erythroid cells, i.e.&nbsp;the cells predecessing erythrocytes in the bone marrow.</p>
</div>
</div>
</div>
<p>We can also evaluate the relative expression of other genes in our dataset, for example, the ones that are most highly expressed. Some very highly expressed genes might point to a technical cause, and we might consider to remove them. Below you will find a simple function to generate a boxplot of relative counts per gene per cell. Load it into your environment and run it on our <code>seu</code> object:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-17_dc19acc6745e476a1908bb4a1daa4013">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>most_expressed_boxplot <span class="ot">&lt;-</span> <span class="cf">function</span>(object, <span class="at">ngenes =</span> <span class="dv">20</span>){</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># matrix of raw counts</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  cts <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">GetAssayData</span>(object, <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">slot =</span> <span class="st">"counts"</span>)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get percentage/cell</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>  cts <span class="ot">&lt;-</span> <span class="fu">t</span>(cts)<span class="sc">/</span><span class="fu">colSums</span>(cts)<span class="sc">*</span><span class="dv">100</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  medians <span class="ot">&lt;-</span> <span class="fu">apply</span>(cts, <span class="dv">2</span>, median)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get top n genes</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>  most_expressed <span class="ot">&lt;-</span> <span class="fu">order</span>(medians, <span class="at">decreasing =</span> T)[ngenes<span class="sc">:</span><span class="dv">1</span>]</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  most_exp_matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>((cts[,most_expressed]))</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># prepare for plotting</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>  most_exp_df <span class="ot">&lt;-</span> <span class="fu">stack</span>(<span class="fu">as.data.frame</span>(most_exp_matrix))</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(most_exp_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"perc_total"</span>, <span class="st">"gene"</span>)</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># boxplot with ggplot2</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>  boxplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(most_exp_df, <span class="fu">aes</span>(<span class="at">x=</span>gene, <span class="at">y=</span>perc_total)) <span class="sc">+</span></span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_flip</span>()</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(boxplot)</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a><span class="fu">most_expressed_boxplot</span>(seu, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="analysis_tools_qc_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>As for most 10X based poly-A enriched single cell datasets, we find a relatively high expression of MALAT1. Many researchers choose to remove it, but it can have biological relevance (e.g.&nbsp;<a href="https://www.nature.com/articles/s41420-020-00383-y">Shaat et al.&nbsp;2021</a>).</p>
</section>
<section id="cell-filtering" class="level3">
<h3 class="anchored" data-anchor-id="cell-filtering">Cell filtering</h3>
<p>Based on the QC process we went through we can come to the following conclusions:</p>
<ul>
<li>There are no cells with very high mitochondrial gene counts.</li>
<li>There are some cells with a hemoglobin and low ribosomal counts, and these are probably erythrocytes.</li>
<li>There are some cells with a very low and very high number of features. These might point to non-informative cells and doublets respectively.</li>
<li>The ‘usual suspect’ MALAT1 sometimes makes up a large part of the counts per cell. As we do not see any other suggestions of dying/stressed cells, we leave it in.</li>
</ul>
<p>In this case, a sensible decision would be to do mild filtering on the number of features per cell and mitochondrial counts. We can leave the possible erythrocytes in for now, and see where they end up later during the dimensionality reduction.</p>
<p>In the M&amp;M of the <a href="https://www.nature.com/articles/s41598-020-64929-x#Sec7">publication</a>, the authors describe that they have used a threshold of &lt; 8% mitochondrial counts and &gt; 200 features per cell. To filter against possible doublets, here, we also filter out cells with &gt; 5000 detected features/cell. Filtering <code>Seurat</code> objects can be done with the <code>subset</code> method for class <code>SeuratObject</code>:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-18_77e41082d5c98b7e8f589abc1fb56918">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">subset</span>(seu, <span class="at">subset =</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">200</span> <span class="sc">&amp;</span> </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>                nFeature_RNA <span class="sc">&lt;</span> <span class="dv">5000</span> <span class="sc">&amp;</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                percent.mito <span class="sc">&lt;</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To evaluate this did the trick we can visualize those parameters again in a violin plot:</p>
<div class="cell" data-hash="analysis_tools_qc_cache/html/unnamed-chunk-19_00ca7ba009739be31d1e9cec87dcf4bd">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">VlnPlot</span>(seu, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>,</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"percent.mito"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="analysis_tools_qc_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>